Make `Products by category/tag/attribute` fallback to the `Product catalog` template

albarin opened the pull request:
This PR aims to make the `Products by Category`, `Products by Tag`, and `Products by Attribute` templates fall back to the `Product Catalog` one.
That means that if neither the theme nor the user has customized one of those three templates, it will look like the Product Catalog one.

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/7648

### User Facing Testing

#### Pre-requisites for testing
- Make sure you are on [this branch](https://github.com/woocommerce/woocommerce/pull/35617) on WC Core if the change is not yet on the latest version of the plugin.
- Make sure you have at least one tag and some products tagged with it.
- Make sure you have at least one Product Attribute with `Enabled Archives` and some products associated with it.
- Make sure you have at least one category and some products on it.
- Make sure you are using the "Twenty Twenty Two" theme (or a blocks theme).

#### `Product by category/tag/attribute` fall back to `Product Catalog` if not customized
1. Make you have no customizations on any template on `https://store.local/wp-admin/site-editor.php?postType=wp_template`.
2. Edit the `Product Catalog` template, add some customization and save.
3. On the front end, go to `/shop` and make sure you see the customization you just did.
4. On the front end, go also to a category, a tag, and an attribute page (like `product-category/clothing/`, `/product-tag/music`, `color/red`, depending on your store configuration)./
5. Check that all 3 pages are using the `Product Catalog` template, meaning you see exactly the same customization you did on the `Product Catalog`.

#### `Product by category/tag/attribute` customizations take priority over `Product Catalog` ones
1. Make you have no customizations on any template on `https://store.local/wp-admin/site-editor.php?postType=wp_template`.
2. Edit the `Product by Attribute` template, add some customization and save.
3. Go back, edit the `Product Catalog` template, add some customization and save.
4. On the front end, go to a product attribute page and make sure you see the customization you just did to `Product by Attribute`.
5. On the front end, go to `/shop`, to a category, and a tag page and check that you see the customization from `Product Catalog`.
6. Clear all customizations and repeat for `Product by Tag` and `Product by Category` templates.

#### `Product by category/tag/attribute` fall back to `Product Catalog` from the theme
1. Make you have no customizations on any template on `https://store.local/wp-admin/site-editor.php?postType=wp_template`.
2. Add a file template called `archive-product.html` to your `templates` folder of your theme with some identifiable changes.
3. On the front end, go to `/shop` and make sure you see the template you just added.
4. On the front end, go also to a category, a tag, and an attribute page (like `product-category/clothing/`, `/product-tag/music`, `color/red`, depending on your store configuration).
5. Check that all 3 pages are also using the `Product Catalog` template from the theme.

#### `Product by category/tag/attribute` from theme take priority over `Product Catalog`
1. Make you have no customizations on any template on `https://store.local/wp-admin/site-editor.php?postType=wp_template`.
2. Add a file template called `archive-product.html` to your `templates` folder of your theme with some identifiable changes.
3. Add a file template called `taxonomy-product_attribute.html` to your `templates` folder of your theme with some identifiable changes.
5. On the front end, go to a product attribute page and make sure it's using the `taxonomy-product_attribute.html` template.
6. On the front end, go to `/shop`, to a category, and a tag page and check that they are using the `archive-product.html`.
7. Repeat for `Product by Tag` (file `taxonomy-product_tag.html`) and `Product by Category` (file `taxonomy-product_cat.html`).

#### Test any other combination of file templates and customized templates in the db you can think of.
In general, the priority should be:
- Custom `Taxonomy` template
- `Taxonomy` template from the theme
- Custom `Product Catalog` template
- `Product Catalog` template from the theme
- `Product Catalog` template from blocks

I made this table 👇 with all the possible combinations of theme and customized templates and the result template I think they should render (could be useful to make some more testing cases):
<img width="988" alt="Screenshot 2022-11-21 at 11 56 13" src="https://user-images.githubusercontent.com/186112/203033261-fd1f9eab-1e53-4858-a049-412b8d13f9c7.png">

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Make the 'Products by Category', 'Products by Tag', and 'Products by Attribute' templates to fallback to the 'Product Catalog' one.

albarin committed:
Add archive-product to the hierarchy

albarin committed:
Fallback to archive-product and get titles from blocks if the templates from theme don't have one

albarin committed:
Rename function

albarin committed:
Fix comment

albarin committed:
Add fallbacks to the db and blocks versions

albarin committed:
Add missing product attribute blockified template

albarin committed:
Update docs

albarin committed:
Add comment and fix comment type on wp_template

albarin committed:
Merge branch 'trunk' into 7648/exploration-fallback-2

albarin committed:
Replate template name if we know the template in blocks

albarin committed:
Add comment and fix linting error

albarin committed:
Fix archive-product template

albarin committed:
Clone the fallback template from db to show them in the template lsit

albarin committed:
Merge branch 'trunk' into 7648/exploration-fallback-2

albarin committed:
Merge branch 'trunk' into 7648/exploration-fallback-2

Aljullu commented on src/Utils/BlockTemplateUtils.php:
```diff
@@ -452,26 +453,84 @@ public static function get_block_template( $id, $template_type ) {
+		}
+
+		foreach ( $db_templates as $template ) {
+			if ( self::template_is_eligible_for_product_archive_fallback( $template_slug )
```
Do we need this check `self::template_is_eligible_for_product_archive_fallback( $template_slug )`? It seems duplicated from line 500, no?

albarin replied:
You are right, thanks!

Aljullu commented on src/Utils/BlockTemplateUtils.php:
```diff
@@ -452,26 +453,84 @@ public static function get_block_template( $id, $template_type ) {
+		$array_filter = array_filter(
+			$db_templates,
+			function ( $template ) use ( $template_slug ) {
+				return self::template_is_eligible_for_product_archive_fallback( $template_slug )
```
Same here, we already returned `false` if `self::template_is_eligible_for_product_archive_fallback( $template_slug )` evaluates to false in line 476-479, no?

Aljullu reviewed (approved):
This is testing great, thanks so much for writing those testing steps with so much detail!

I left a couple of questions inline in the code, but besides that, LGTM!

albarin committed:
Return the fallback template when querying for a single template

albarin committed:
Remove unneeded condition

albarin committed:
Merge branch 'trunk' into 7648/exploration-fallback-2

albarin committed:
Use wp function instead of gutenberg one

albarin committed:
Fix tests

albarin committed:
Merge branch 'trunk' into 7648/exploration-fallback-2

albarin committed:
Fix tests on Product Catalog templates

It was checking a single product for the customization, but it should check the /shop page

albarin commented:
@Aljullu I've made some more changes, in case you want to review:
There was an issue when editing the template directly from the specific URL (`wp-admin/site-editor.php?postType=wp_template&postId=woocommerce%2Fwoocommerce%2F%2Farchive-product`), instead of from the list of templates, it wasn't loading the fallback template.
I've fixed it with the `pre_get_block_template` hook.
```
add_filter( 'pre_get_block_template', array( $this, 'get_block_template_fallback' ), 10, 3 );
```

There are a few tests failing right now because of an issue that is being fixed here: https://github.com/WordPress/gutenberg/pull/45992, once that is merged and we update the version of the `e2e-test-utils` they should pass.
They are passing locally when I test with the change.

albarin committed:
Merge branch 'trunk' into 7648/exploration-fallback-2

Aljullu commented on tests/e2e/specs/backend/site-editing-templates.test.js:
```diff
@@ -155,6 +155,11 @@ describe( 'Store Editing Templates', () => {
 	} );
```
Now that this `beforeAll()` is inside each inner clause, do you think we can remove it from the top?

albarin replied:
Yes, I will change that and disable the tests. 👍

dinhtungdu replied:
> Do you think we should disable those test until we can update e2e-test-utils and get them pass?

> Yes, I will change that and disable the tests. 👍

Did we log this somewhere (eg: an issue) so we can remember to 'unskip' those tests again when we update `e2e-test-utils`?

albarin replied:
Yep, I created an issue this morning https://github.com/woocommerce/woocommerce-blocks/issues/7744

Aljullu reviewed (approved):
Thanks for investigating the failing tests, @albarin! I did a quick round of testing and everything seems to be working well, and code changes look good.

> There are a few tests failing right now because of an issue that is being fixed here: https://github.com/WordPress/gutenberg/pull/45992, once that is merged and we update the version of the `e2e-test-utils` they should pass.

Do you think we should disable those test until we can update `e2e-test-utils` and get them pass?

albarin committed:
Disable tests related with deleteAllTemplates function
